### PR TITLE
Add Razer Naga Pro Wired/Wireless Support 

### LINKED ISSUE #1247: Support for Razer Naga Pro
### Device Info

* Name: Razer Naga Pro
* Product Number: RZ01-03420100
* Model Number: RZ01-0342

### Device Information

```
Bus 001 Device 011: ID 1532:0090 Razer USA, Ltd Razer Naga Pro
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0090 
  bcdDevice            1.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 3 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          4 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     325
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
```

```
lrwxrwxrwx 1 root root   10 Sep  9 12:36 usb-RAZER_Razer_Mouse_Dock-event-if01 -> ../event30
lrwxrwxrwx 1 root root   10 Sep  9 12:36 usb-RAZER_Razer_Mouse_Dock-event-mouse -> ../event26
lrwxrwxrwx 1 root root   10 Sep  9 12:36 usb-RAZER_Razer_Mouse_Dock-if01-event-kbd -> ../event27
lrwxrwxrwx 1 root root   10 Sep  9 12:36 usb-RAZER_Razer_Mouse_Dock-if02-event-kbd -> ../event31
lrwxrwxrwx 1 root root    9 Sep  9 12:36 usb-RAZER_Razer_Mouse_Dock-mouse -> ../mouse0
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-event-if01 -> ../event263
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-event-if03 -> ../event268
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-event-mouse -> ../event257
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-if01-event-kbd -> ../event258
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-if02-event-kbd -> ../event266
lrwxrwxrwx 1 root root    9 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-mouse -> ../mouse1
```

### PR DESCRIPTION
**Pull request by AliceGrey**
Add Razer Naga Pro Support

Also Fixes #1247 (once completed)

So far what's working:

    DPI
    Polling
    Scroll Wheel LED
    Logo LED
    Pretending the left thumbgrid is the backlight LED
    Battery monitoring
    Wireless

What's not working:

    Addressing just the left thumbgrid LED independently
    Spectrum, Reactive, Breath for left thumbgrid
    Tilt wheel?

### CODE REVIEW

**Comment by AliceGrey** [comment]
~~@lah7 For some reason while connected wirelessly the DPI settings do not show for this device in Polychromatic. They do however show when using a wired connection. The wireless class inherits all the properties from the wired class. Any ideas on what may be causing this?~~

@z3ntu Would you have any ideas on why 'get_left_brightness', 'set_left_brightness' aren't working for this device before my commit to change it to using the backlight? I'm able to address the thumbgrid independently in Polychromatic with the matrix coordinates 2,0.

**Comment by nightsky30** [comment]
> ```
> Wireless (But DPI settings don't show in Polychromatic for some reason?)
> ```
This might be a sign of a timing issue for the wireless dongle. There have been a few wireless devices which would only return values such as DPI/Brightness intermittently, or not at all until the wait times were adjusted for the device e.g. the Lancehead Wireless, Mamba Wireless, Atheris, and some others. You might try adding a case statement for this mouse:
https://github.com/openrazer/openrazer/blob/45c7d0cd8ab2466a8ceb1e6024c150addec11bf6/driver/razermouse_driver.c#L51-L82

**Comment by AliceGrey** [comment]
@nightsky30 that did the trick! Thank you for the tip. You wouldn't happen to be familiar with the left thumbgrid issue I mentioned above?

**Comment by nightsky30** [comment]
Glad that worked! :smile: You're welcome.

Hmmm. So it sounds like you have had success somehow with the backlight LED value? Like do the backlight effects and backlight brightness affect ALL LEDs, or just the Left Thumbgrid LEDs? If it only affects the thumbgrid, I'm guessing that might just be the backlight.

- ~Just as a test, have you tried substituting the `LEFT_SIDE_LED` constant in one of the left wrapper methods with the `BACKLIGHT_LED`? And does that only set the effect for the thumbgrid?~

https://github.com/openrazer/openrazer/blob/1d80f1bdc4a25795de730acf07f19cc5b2c5e047/driver/razermouse_driver.c#L3468-L3476

**EDIT**: Well, the above wouldn't work on it's own if the class for the mouse doesn't have the Left specific capability methods in `mouse.py`. And you stated it seems to be responding with the Backlight LED value....Maybe just try what's below regarding the general "Backlight" effect methods. You already have those methods in the class:

Something I noticed while scanning through the changes was the general "Backlight" methods such as `razer_attr_write_mode_spectrum` are currently making use of `razer_chroma_mouse_extended_matrix_effect_spectrum` or `razer_chroma_standard_matrix_effect_spectrum` when setting the effects for the other Naga mice. BUT there are also non-mouse specific extended effect versions of each effect like `razer_chroma_extended_matrix_effect_spectrum`. Other new mice have used these extended matrix effects.

- Have you tried `razer_chroma_extended_matrix_effect_spectrum` with the `BACKLIGHT_LED` (and transaction ID `0x1f`, of course)?

https://github.com/openrazer/openrazer/blob/1d80f1bdc4a25795de730acf07f19cc5b2c5e047/driver/razermouse_driver.c#L706-L737

If those fail the only other ideas I have at the moment are perhaps:
- Switching to a different device mode (like mode 3 vs 0) before setting the backlight LED effect
- Or maybe the thumbgrid has a different LED value of which we aren't aware?

**Comment by AliceGrey** [comment]
@nightsky30 

The backlight sets all LED's to one mode / color for the naga pro. It doesn't provide control over just the left thumbpad unfortunately.

**Comment by nightsky30** [comment]
Ah, ok. So since you said you can independently modify the thumbgrid with the LED at (2,0), then there might be a unique Thumbgrid LED value different from the Left Side LED other mice use.

Not sure there is an easy way to identify those values other than trial and error. Maybe build a loop into one of the set effect methods that prints the LED value it's trying and just have it loop through and send a range of values? Maybe the daemon has a way of exposing the values of LEDs it detects?

https://github.com/openrazer/openrazer/blob/64d589d0d0426007360991f582e97e77d51a4f41/driver/razercommon.h#L53-L68

**Comment by AliceGrey** [comment]
Hitting another roadblock with this. For some reason after updating to the latest 5.15 kernel the battery level is not reporting :(

Polychromatic:
`FileNotFoundError: [Errno 2] No such file or directory: ’/sys/devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.3/usb3/3-1/3-1:1.0/0003:1532:0090.0002/charge_level’`

**Comment by AliceGrey** [comment]
Wired input also no longer works

Razer Mount log:
```

Dec 04 09:40:32 fenrir razer_mount[7435]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7436]: Device_ID 1-1
Dec 04 09:40:32 fenrir razer_mount[7444]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7447]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7448]: Device_ID 1-1:1.0
Dec 04 09:40:32 fenrir razer_mount[7449]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7450]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7451]: Device_ID 1-1:1.2
Dec 04 09:40:32 fenrir razer_mount[7452]: Device_ID 1-1:1.1
Dec 04 09:40:32 fenrir razer_mount[7453]: Device_ID 1-1:1.3
Dec 04 09:40:32 fenrir razer_mount[7457]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7459]: Device_ID 0003:1532:008F.002A
Dec 04 09:40:32 fenrir razer_mount[7460]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7461]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7462]: Driver razermouse
Dec 04 09:40:32 fenrir razer_mount[7463]: Unbinding 0003:1532:008F.002A from hid-generic
Dec 04 09:40:32 fenrir razer_mount[7465]: Device_ID 0003:1532:008F.002C
Dec 04 09:40:32 fenrir razer_mount[7464]: Device_ID 0003:1532:008F.002B
Dec 04 09:40:32 fenrir razer_mount[7466]: Device_ID 0003:1532:008F.002D
Dec 04 09:40:32 fenrir razer_mount[7467]: Unbinding 0003:1532:008F.002C from hid-generic
Dec 04 09:40:32 fenrir razer_mount[7468]: Unbinding 0003:1532:008F.002B from hid-generic
Dec 04 09:40:32 fenrir razer_mount[7469]: Unbinding 0003:1532:008F.002D from hid-generic
Dec 04 09:40:32 fenrir razer_mount[7470]: Binding 0003:1532:008F.002D to razermouse
Dec 04 09:40:32 fenrir systemd-udevd[7438]: 0003:1532:008F.002D: Process 'razer_mount razermouse 0003:1532:008F.002D' failed with exit code 1.
Dec 04 09:40:32 fenrir razer_mount[7473]: Binding 0003:1532:008F.002A to razermouse
Dec 04 09:40:32 fenrir systemd-udevd[7421]: 0003:1532:008F.002A: Process 'razer_mount razermouse 0003:1532:008F.002A' failed with exit code 1.
Dec 04 09:40:32 fenrir razer_mount[7475]: Binding 0003:1532:008F.002B to razermouse
Dec 04 09:40:32 fenrir systemd-udevd[7437]: 0003:1532:008F.002B: Process 'razer_mount razermouse 0003:1532:008F.002B' failed with exit code 1.
Dec 04 09:40:33 fenrir razer_mount[7477]: Binding 0003:1532:008F.002C to razermouse
Dec 04 09:40:33 fenrir systemd-udevd[7440]: 0003:1532:008F.002C: Process 'razer_mount razermouse 0003:1532:008F.002C' failed with exit code 1.

```

Dmesg
```

[  596.611582] usb 1-1: new full-speed USB device number 5 using xhci_hcd
[  596.824308] usb 1-1: New USB device found, idVendor=1532, idProduct=008f, bcdDevice= 1.00
[  596.824309] usb 1-1: New USB device strings: Mfr=0, Product=2, SerialNumber=3
[  596.824310] usb 1-1: Product: Razer Naga Pro
[  596.824311] usb 1-1: SerialNumber: 000000000000
[  596.843370] input: Razer Naga Pro as /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.0/0003:1532:008F.002A/input/input114
[  596.843444] hid-generic 0003:1532:008F.002A: input,hidraw1: USB HID v1.11 Mouse [Razer Naga Pro] on usb-0000:05:00.1-1/input0
[  596.851459] input: Razer Naga Pro as /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.2/0003:1532:008F.002B/input/input115
[  596.921725] hid-generic 0003:1532:008F.002B: input,hidraw2: USB HID v1.11 Keyboard [Razer Naga Pro] on usb-0000:05:00.1-1/input2
[  596.928875] input: Razer Naga Pro Keyboard as /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C/input/input116
[  596.991694] input: Razer Naga Pro Consumer Control as /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C/input/input117
[  596.991723] input: Razer Naga Pro System Control as /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C/input/input118
[  596.991748] input: Razer Naga Pro as /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C/input/input119
[  596.991790] hid-generic 0003:1532:008F.002C: input,hidraw3: USB HID v1.11 Keyboard [Razer Naga Pro] on usb-0000:05:00.1-1/input1
[  596.998462] hid-generic 0003:1532:008F.002D: hiddev96,hidraw4: USB HID v1.11 Device [Razer Naga Pro] on usb-0000:05:00.1-1/input3
```
Razer log
```
2021-12-04 09:40:28 | razer                          | DEBUG    | Device event [remove]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.0/0003:1532:008F.0026
2021-12-04 09:40:28 | razer                          | DEBUG    | Device event [remove]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.2/0003:1532:008F.0027
2021-12-04 09:40:28 | razer                          | DEBUG    | Device event [remove]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.0028
2021-12-04 09:40:28 | razer                          | DEBUG    | Device event [remove]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.3/0003:1532:008F.0029
2021-12-04 09:40:32 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.3/0003:1532:008F.002D
2021-12-04 09:40:32 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.3/0003:1532:008F.002D
2021-12-04 09:40:32 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.0/0003:1532:008F.002A
2021-12-04 09:40:32 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.0/0003:1532:008F.002A
2021-12-04 09:40:32 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.2/0003:1532:008F.002B
2021-12-04 09:40:32 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.2/0003:1532:008F.002B
2021-12-04 09:40:33 | razer                          | DEBUG    | Device event [add]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C
2021-12-04 09:40:33 | razer                          | DEBUG    | Device event [bind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C
2021-12-04 09:40:33 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.3/0003:1532:008F.002D
2021-12-04 09:40:33 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.0/0003:1532:008F.002A
2021-12-04 09:40:33 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:05:00.1/usb1/1-1/1-1:1.1/0003:1532:008F.002C
2021-12-04 09:40:33 | razer                          | DEBUG    | Device event [unbind]: /devices/pci0000:00/0000:00:01.2/0000:01:00.0/0000:02:08.0/0000:
```05:00.1/usb1/1-1/1-1:1.2/0003:1532:008F.002B

File: driver/razermouse_driver.c [hunk 1]
```diff
@@ -5143,6 +5366,46 @@ static void razer_mouse_disconnect(struct hid_device *hdev)
...
             break;
 
+        case USB_DEVICE_ID_RAZER_NAGA_PRO_WIRELESS:
+            device_remove_file(&hdev->dev, &dev_attr_charge_effect);
+            device_remove_file(&hdev->dev, &dev_attr_charge_colour);
+            fallthrough;
+        case USB_DEVICE_ID_RAZER_NAGA_PRO_WIRED:
+            device_remove_file(&hdev->dev, &dev_attr_dpi);
+            device_remove_file(&hdev->dev, &dev_attr_poll_rate);
+
```
Comment (nightsky30) [inline driver/razermouse_driver.c]: Looks like `device_remove_file(&hdev->dev, &dev_attr_dpi_stages);` should be added here to match the section with device file creation

**Comment by nightsky30** [comment]
Does this happen every time, or just sometimes? What happens if you replug the mouse/wireless dongle?

The first error seems to indicate the device file wasn't created or at least not created where it was expected, but I didn't see anything really missing in the PR.

I thought the number of event files and interfaces for this device was interesting. Some accessories have needed a special check to force the use of one interface while ignoring others, but I believe the errors they were seeing in their logs were different. I'm wondering if this might be the same issue.

<details>
<summary>
ls -lah /dev/input/by-id/
</summary>

lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-event-if01 -> ../event263
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-event-if03 -> ../event268
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-event-mouse -> ../event257
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-if01-event-kbd -> ../event258
lrwxrwxrwx 1 root root   11 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-if02-event-kbd -> ../event266
lrwxrwxrwx 1 root root    9 Sep  9 12:36 usb-Razer_Razer_Naga_Pro_000000000000-mouse -> ../mouse1

</details>
<details>
<summary>
lsusb -d 1532: -v
</summary>

Bus 001 Device 011: ID 1532:0090 Razer USA, Ltd Razer Naga Pro
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0090 
  bcdDevice            1.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 3 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          4 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              100mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     325
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               1
</details>

Example of what they did in the accessory driver:
https://github.com/openrazer/openrazer/blob/490f1f3b0c887e702b43f343fbdc9c845b2c777b/driver/razeraccessory_driver.c#L1676-L1697
https://github.com/openrazer/openrazer/blob/490f1f3b0c887e702b43f343fbdc9c845b2c777b/driver/razeraccessory_driver.c#L2110

**Comment by Olyve** [comment]
I'm new to this but would love to see some support for the device so please let me know if there is some way I can help!

**Comment by rastafaninplakeibol** [comment]
Hello @AliceGrey, how can I try your branch? I tried cloning your repo and then make/make install_i_know_what_im_doing but the drivers did not work with the polychromatic GUI (it says that it couldn't detect any backend of openrazer). How did you manage to install and test your version of these drivers?

**Comment by z3ntu** [comment]
@rastafaninplakeibol https://github.com/openrazer/openrazer/wiki/Building-a-package

And please don't use `install_i_know_what_im_doing`, it's named like that for a reason :facepalm:

**Comment by rastafaninplakeibol** [comment]
ok first of all thanks @z3ntu, i managed to build everything and it seems to work now with the different GUIs, but the mouse does not pop up as a supported device even if I can see in dmesg that the kernel indeed recognize the device in all the 3 different configurations (wired/BT/dongle). @AliceGrey did you write also a custom cli to use your drivers? How can I help testing it?

**Comment by CrazyM0nkey** [comment]
Just installed AliceGrey:razer-naga-pro and so far the dock(as Razer Mouse Dock) and the mouse(as Razer Naga Pro (Wireless)) are recognized.
I can change the DPI and the LED's.
Installed on Manjaro, kernel 5.16.2-1

**Comment by z3ntu** [comment]
@CrazyM0nkey Do you see issues like described above in https://github.com/openrazer/openrazer/pull/1692#issuecomment-986044463 ?

@AliceGrey Any update from your side?

**Comment by Keralin** [comment]
I installed yesterday this branch with kernel `5.13.0-30-generic` on Ubuntu and everything worked as expected with also de battery level of the mouse when wireless.

**Comment by ciferkey** [comment]
Just installed and tested the branch on 
```
$ uname -a
Linux ubuntu 5.13.0-19-generic #19-Ubuntu SMP Thu Oct 7 21:58:00 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```
It detects the mouse perfectly:
image

Is there anything specific I can test / logs I can pull to help move this branch forwards?

**Comment by CrazyM0nkey** [comment]
> 

No, I have no issue

**Comment by CrazyM0nkey** [comment]
> 
@z3ntu No, I have no issues... Works with Polychromatic as well as  Snake.

**Review by z3ntu — APPROVED** [review]
Thanks everybody for testing, and @AliceGrey for the code!

LGTM

**Comment by LEOMODE** [comment]
Is there a way for a dummy like me to just use the package installation in the Mac OS? :(

**Comment by c0d3m0nky** [comment]
I've got it seeing my Naga Pro after adding the daily ppa, but all I have available is lighting and DPI. This is my first time using OpenRazer, my understanding is that I could control button profiles like on Windows. Is that not the case, or is more work needed to implement that? I've tried with RazerGenie and Snake.

**Pull request merged**